ocr-gcp-cloud-vision/python/services/ocr.py

```diff
@@ -0,0 +1,256 @@
+                "city": ctx.req.json["city"],
+            },
+        )
+
+        # Return an HTTP 200 OK message stating that the creation of the user profile
+        # happened successfully.
+        ctx.res.body = {
+            "msg": f"Profile with id '{user_profile_id}' created successfully."
+        }
+        # Return `None` to break the flow and return the intended response.
+        return
+    except Exception:
+        # Return an HTTP 500 Internal Server Error response in case the Key-Value Store
+        # specific `set` operation didn't succeed.
+        ctx.res.status = 500
+        ctx.res.body = {
+            "msg": "An Internal Server Error happened. User profile creation failed."
+        }
+        # Return `None` to break the flow and return the intended response.
+        return
+
+
+@ocr_api.get("/profiles/:id")
+async def get_user_profile(ctx: HttpContext) -> None:
+    """
+    A GET API to retrieve a user's profile details.
+
+    Args:
+        ctx (HttpContext): HTTP Request and Response context for the API.
+    """
+    # Get the profile id of the user from the HTTP Request.
+    user_profile_id = ctx.req.params["id"]
+
+    # Retrieve the details for the user's profile.
+    try:
+        profile_details = await user_profiles_kv.get(user_profile_id)
+        ctx.res.headers["Content-Type"] = "application/json"
+        ctx.res.body = f"{json.dumps(profile_details)}"
+    except Exception:
+        # Return an HTTP 500 Internal Server Error response in case the Key-Value Store
+        # specific `get` operation didn't succeed.
+        ctx.res.status = 500
+        ctx.res.body = {
+            "msg": "An Internal Server Error happened. User profile couldn't be fetched or doesn't exist."
+        }
+        return
+
+
+@ocr_api.get("/profiles/:id/image/upload")
+async def upload_image(ctx: HttpContext) -> None:
+    """
+    A GET API to retrieve a Signed URL for uploading
+    the image.
+
+    Args:
+        ctx (HttpContext): HTTP Request and Response context for the API.
+    """
+    # Get the profile id of the user from the HTTP Request.
+    user_profile_id = ctx.req.params["id"]
+
+    # Generate a Signed URL which is valid for a defined duration.
+    # It should be noted that having user identifiable information in
+    # the URL or HTTP Payload is not a good security practice and could expose the APIs
+    # to OWASP API and Web Application exploits. The shown Signed URL
+    # generation scheme is just for demonstration purposes.
+    photo_url = await photos.file(f"images/{user_profile_id}/photo.png").upload_url(
+        expiry=timedelta(seconds=(3600))
+    )
+
+    # Set the expiry duration so that the same can be passed in the HTTP Response
+    # headers.
+    expires = (datetime.now(UTC) + timedelta(seconds=(3600))).strftime(
+        "%a, %d %b %Y %H:%M:%S GMT"
+    )
+    ctx.res.headers["Expires"] = expires
+
+    # Set the HTTP Response body.
+    ctx.res.body = photo_url
```

Comment:
Making this a redirect could simplify how this is handled on the client side, unless the idea is to demonstrate the signed URL result.

Reply:
The final example on this page shows how to return a signed URL as a redirect https://nitric.io/docs/reference/python/storage/bucket-file-downloadurl

Reply:
That's a fair suggestion, Tim.
From a client experience perspective, redirection flow is better.
I will adjust the code accordingly.